### PR TITLE
working on linear modeling

### PR DESCRIPTION
**Pull request by Jay4869**
1. make linear modeling for `bold.nii.gz` and `task#_run#_conv#`
2. return the beta_hat and create a `beta` folder to store all the beta we are computing
3. create a `convo` folder to store all `task#_run#_conv#` image and text files

Question: how can we test on our beta to know which brain area is significant with experiment signal

### CODE REVIEW

**Comment by Jay4869** [comment]
I delete the branch `convo`, but right now everything is on `modeling` branch
Please check this

**Comment by lizeyuyuz** [comment]
@Jay4869  Once we get the estimated betas, we can do t-test to check which ones are significant. To get the standard errors of the beta_hats, we would have to find the variance-covariance matrix.

**Comment by Jay4869** [comment]
I have updated all the images (convolving response) what I produced the so far. At this point, the code is fulling worked, and if no more concern, please merge to upstream and everyone can update to the latest version

**Comment by Jay4869** [comment]
Check two type images, comment which one you prefer to put on report!

**Comment by Jay4869** [comment]
debugging on linear modeling and produced one slice beta

File: code/utils/basic_modeling.py [hunk 1]
```diff
@@ -2,46 +2,73 @@
...
+    betas = npl.pinv(design).dot(data_2d.T) #(2, 147456)
+    betas_4d = np.reshape(betas.T, data.shape[:-1] + (-1,)) #(64, 64, 36, 2)
+
+    return (design, data_2d, betas, betas_4d)   
 
-    return betas_4d
+def RSE(X,Y, betas_hat): #input 2D X, Y and betas_hat
+    Y_hat = X.dot(betas_hat)
+    res = Y - Y_hat.T
+    RSS = np.sum(res ** 2)
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: axis = 0

File: code/utils/basic_modeling.py [hunk 2]
```diff
@@ -2,46 +2,73 @@
...
+    df = X.shape[0] - npl.matrix_rank(X)
+    MRSS = RSS / df
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: Note: axis = 0

File: data/convo/task001_run001_conv005.txt [hunk 3]
```diff
@@ -0,0 +1,137 @@
+0.000000000000000000e+00
```
Comment (lizeyuyuz) [inline data/convo/task001_run001_conv005.txt]: We should be only using cond005, can you remove other conv00#.txt files?

**Comment by Jay4869** [comment]
@lizeyuyuz I have already removed rest conv#, only keep conv005

File: code/utils/basic_modeling.py [hunk 4]
```diff
@@ -3,45 +3,46 @@
...
     design = np.ones((len(convolved), 2))
     design[:, 1] = convolved
     data_2d = np.reshape(data, (-1, data.shape[-1]))
-    betas = npl.pinv(design).dot(data_2d.T)
-    betas_4d = np.reshape(betas.T, img.shape[:-1] + (-1,))
+    betas = npl.pinv(design).dot(data_2d.T) #(2, 147456)
+    betas_4d = np.reshape(betas.T, data.shape[:-1] + (-1,)) #(64, 64, 36, 2)
 
-    return betas_4d
+    return (betas, betas_4d)
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: Your betes outpt txt file is all zeros. Need to check this function!

File: code/utils/basic_modeling.py [hunk 5]
```diff
@@ -2,46 +2,82 @@
...
     design = np.ones((len(convolved), 2))
     design[:, 1] = convolved
-    data_2d = np.reshape(data, (-1, data.shape[-1]))
-    betas = npl.pinv(design).dot(data_2d.T)
-    betas_4d = np.reshape(betas.T, img.shape[:-1] + (-1,))
+    data_2d = np.reshape(data, (data.shape[-1], -1))
+    betas = npl.pinv(design).dot(data_2d) #(2, 147456)
+    betas_4d = np.reshape(betas.T, data.shape[:-1] + (-1,)) #(64, 64, 36, 2)
+    print(design.shape, data_2d.shape, betas.shape)
+    return (design, data_2d, betas, betas_4d)
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: @Jay4869 The betas output txt file contains only zeros. Check this function!

File: code/utils/basic_modeling.py [hunk 6]
```diff
@@ -2,46 +2,82 @@
...
+    df = X.shape[0] - npl.matrix_rank(X)
+    MRSS = RSS / df
 
-    return betas_4d
+    return (MRSS, df) 
+
+def hypothesis(betas_hat, v_cov, df):
+    t_stat = np.array([])
+    for i in v_cov:
+        t_stat = np.append(t_stat, betas_hat[:, 1] / np.sqrt(i[1,1]))
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: I think this t_stat is wrong, your v_cov is always 2 by 2, so you don't need to use a for loop.

File: code/utils/basic_modeling.py [hunk 7]
```diff
@@ -2,46 +2,82 @@
...
+    design_mat, data_2d, betas_hat, betas_hat_4d = reg_voxels_4d(data, convolved)
+    np.savetxt('../../data/beta/' + get_name + '.txt', betas_hat, newline='\r\n')
+    plt.imshow(betas_hat_4d[:, :, 15, 1])
+    plt.savefig('../../data/beta/task001_run001_conv005.png')
+    #plt.show()
+    #get residual standard error
+    s2, df = RSE(design_mat, data_2d, betas_hat)
+    
+    #get estimator of variance of betas_hat   
+    beta_cov = np.array([])
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: I think this t_stat is wrong, your v_cov is always 2 by 2, so you don't need to use a for loop.

File: code/utils/basic_modeling.py [hunk 8]
```diff
@@ -2,46 +2,82 @@
...
+    for i in s2:
+        beta_cov = np.append(beta_cov, npl.inv(design_mat.T.dot(design_mat)))
+    beta_cov = beta_cov.reshape(-1,2,2)
+
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: I think this t_stat is wrong, your v_cov is always 2 by 2, so you don't need to use a for loop.

File: code/utils/basic_modeling.py [hunk 9]
```diff
@@ -2,46 +2,82 @@
...
+    print(design.shape, data_2d.shape, betas.shape)
+    return (design, data_2d, betas, betas_4d)   
+
+def RSE(X,Y, betas_hat): #input 2D of X, Y and betas_hat
+    Y_hat = X.dot(betas_hat)
+    res = Y - Y_hat
+    RSS = np.sum(res ** 2, 0)
+    print(RSS.shape)
+    df = X.shape[0] - npl.matrix_rank(X)
+    MRSS = RSS / df
```
Comment (lizeyuyuz) [inline code/utils/basic_modeling.py]: I've checked code from here above, they are all correct, the dimensions are right, and the outputs are reasonable.

**Comment by lizeyuyuz** [comment]
1. Note: v_cov = npl.inv(X.T.dot(X)) (the covariance matrix only depends on design matrix, so it's values are the same for each voxel for a given task)
2. Note: betas_hat.shape = (2, 147456) and so betas_hat.shape(1) = 147456

You should have 147456 t-stats and 147456 p-values, so your dimension is off. 

To test beta_1, 
1. We want all the betas_hat_one for each voxel, so beta_hat_one values should be betas_hat[1, : ] and so betas_hat[1, : ].shape = (147456,)
2. And the corresponding variance for beta_hat_one is the (1,1) entry of the v_cov matrix.

Code modification suggestion:
t_stat = betas_hat[1, : ] / np.sqrt(v_cov[1,1])

**Comment by amandazhang** [comment]
@lizeyuyuz What do you mean by pointing the first statement? Is that the reason you change the inside function of getting t_stat by using v_cov only instead of the original one?

**Comment by lizeyuyuz** [comment]
The s2 for loop doesn't look right ...

Code modification suggestion:
```
    beta_cov = np.zeros(shape=(len(s2),))
    for i in range(0,len(s2)):
        beta_cov[i] = s2[i] * npl.inv(design_mat.T.dot(design_mat))[1,1]

```

**Comment by amandazhang** [comment]
Your code's computing beta_cov which is the input of Jay's written function 'hyphothesis'? Thus what's the input then?

**Comment by amandazhang** [comment]
for i in s2:
        beta_cov = np.append(beta_cov, npl.inv(design_mat.T.dot(design_mat)))
add i in the for loop. And then try again.

**Comment by lizeyuyuz** [comment]
@Jay4869 @amandazhang We are getting the same thing, but I think that appending things to array in for loop takes a long time, but inserting items like the one I did took much less time =) Good work everyone!

**Comment by Jay4869** [comment]
@lizeyuyuz  Great job. Have you get p-value? Just for comparing, I got 0.8 weird still!

**Comment by lizeyuyuz** [comment]
@Jay4869 Yes the first few p-values are around 0.8, but did you get only a single p-value?

**Comment by Jay4869** [comment]
No, I got a lot of p-value, but I did not check the smallest one! I will give you answer tmr morning!

**Comment by Jay4869** [comment]
@lizeyuyuz @amandazhang  I tried to use index to implement instead of using `np.append` but I don't know somehow does not work. Therefore, I still keep `np.append` and produced `beta.txt`, `T-value.txt`, 'P-value.txt' and images

**Comment by amandazhang** [comment]
@Jay4869 But can you still keep the code that you made change on but wouldn't work? I could get a chance to see them when we meet today and try debugging them.

**Comment by Jay4869** [comment]
@now the version on the branch, everything is worked!

**Comment by amandazhang** [comment]
To get work done I will merge this to our origin master, but we would still modify this file in the future to make optimized work.

**Pull request merged**